### ISSUE TITLE
[Bug] ttnn.typecast(float -> int32) saturates positive overflow to INT32_MIN, inverting the sign on 19.34% of the float32 domain

### ISSUE THREAD

**Issue by seonglae** [tenstorrent/tt-metal#55933]
### Describe the bug

`ttnn.typecast(t, ttnn.int32)` returns `-2147483648` for every float input at or above `2^31`, where this repository's explicit golden clamps to `+2147483647`, and returns `-2147483648` for NaN where torch gives `0`. Negative inputs came back correct against torch at every value tested, though not against this repository's own golden, which is the second half of the problem and is settled below. Only the positive side is broken outright, and it is broken by a sign inversion rather than a precision loss: the largest representable positive float becomes the most negative `int32`. That is wrong across **19.34% of the float32 domain**.

### Root cause

The kernel keeps a single saturation constant for both signs. `ckernel_sfpu_typecast.h:194` loads `0x80000000` and says so in its own comment:

```cpp
// result = INT_MIN
TTI_SFPLOADI(p_sfpu::LREG1, sfpi::SFPLOADI_MOD0_FLOATB, 0x8000);
```

Line 196 then narrows the lane mask to `exp < 31`, so every lane with `|x| >= 2^31` keeps that constant untouched, and the two's-complement negate at lines 206-209 fires only where `in < 0`:

```cpp
// LaneEnabled = in < 0
TTI_SFPSETCC(0, p_sfpu::LREG0, 0, sfpi::SFPSETCC_MOD1_LREG_LT0);
// result = -result (two's complement)
TTI_SFPIADD(0, p_sfpu::LCONST_0, p_sfpu::LREG1, sfpi::SFPIADD_MOD1_ARG_2SCOMP_LREG_DST | sfpi::SFPIADD_MOD1_CC_NONE);
```

Since `-INT_MIN == INT_MIN`, a positive-overflow lane has no path to `INT32_MAX` at all.

### Repro

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)
x = torch.tensor([[2147483520.0, 2147483648.0, 3e9, 3.4e38,
                   float("inf"), float("nan"), -2147483648.0, float("-inf")]],
                 dtype=torch.float32)
t = ttnn.from_torch(x, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
print(ttnn.to_torch(ttnn.typecast(t, ttnn.int32)))
print(x.to(torch.int32))
```

### Observed vs expected

float32 input, int32 output, against `torch.Tensor.to(torch.int32)`:

| input | got | want |
|---|---|---|
| 2147483520.0 | 2147483520 | 2147483520 |
| **2147483648.0 (2^31)** | **-2147483648** | 2147483647 |
| **3e9** | **-2147483648** | 2147483647 |
| **3.4e38** | **-2147483648** | 2147483647 |
| **+inf** | **-2147483648** | 2147483647 |
| **NaN** | **-2147483648** | 0 |
| -2147483648.0 | -2147483648 | -2147483648 |
| -inf | -2147483648 | -2147483648 |
| -100.0 | -100 | -100 |
| -2000000000.0 | -2000000000 | -2000000000 |
| -1.0 | -1 | -1 |
| 100.0 | 100 | 100 |

The last four rows are there because they are the first thing worth doubting: the negative and in-range paths are correct, so this is a one-sided defect rather than a broken conversion. `2147483520.0` is the largest float below `2^31` and it is also correct, which puts the boundary exactly where the exponent mask below says it is.

**The two goldens disagree with each other about `int32`, and neither matches the kernel.** The semantics were never pinned down, which is why this survived.

The tt-llk golden clamps explicitly, `golden_generators.py:1951`, and its comment concedes the hardware already needed special handling on the other end:

```python
if output_format == DataFormat.Int32:
    # +1 on the min: hardware uses sign-magnitude representation.
    return torch.clamp(values, -(2**31 - 1), 2**31 - 1).to(out_torch)
```

The ttnn golden has **no `int32` arm at all**. `_typecast_golden_function` in `ttnn/ttnn/operations/core.py` handles `uint8` at `:608-613`, `uint16` at `:615-625` and `uint32` at `:627-629`, each clamping through `int64` first, which makes those three architecture-independent. `int32` reaches none of them and falls through to `return input_tensor.to(...)`.

That fallthrough is itself architecture-dependent, so I am not citing it as agreeing with the tt-llk golden. `torch.Tensor.to(torch.int32)` out of range is a `static_cast`; I measured it saturating to `2147483647` on aarch64 and have **not** measured it on x86_64, where `cvttss2si` yields `INT_MIN`. If torch inherits that, this golden returns different values on the two hosts and silently agrees with the broken kernel on one of them.

The function's own header comment shows this is known territory: *"Integer conversion semantics depend on source dtype/placement and target architecture"*, and *"Integer conversion differs across host, Wormhole/Blackhole, and Quasar paths"*. They wrote all three arms for `uint8` and `uint16` and left `int32`'s saturation direction to a host cast.

**So the ask is two things: fix the kernel's positive saturation, and pin the `int32` golden the way the other three are pinned.**

### Affected input range

Counting float32 bit patterns whose correct `int32` value is representable but whose result is `INT32_MIN`:

| dtype | wrong | of the domain | affected range |
|---|---|---|---|
| float32 | 830,472,191 | **19.34%** | `x >= 2^31`, `+inf`, and all 16,777,214 NaN patterns |
| bfloat16 | 12,671 | **19.33%** | same |

A uniform random bit-pattern sweep measured 190 divergences in 1,000 draws: 188 positive overflow, 2 NaN.

bfloat16 as the source dtype is worse, because `2147483520.0` already rounds up to `2^31` in bf16, so inputs that are in range as float32 diverge once stored as bfloat16.

### Which architectures

Blackhole measured. The same `// result = INT_MIN` construct is at line 206 of the Wormhole kernel, so I expect the same behaviour there, but I have not measured it and am not claiming it.

Four separate LUT entries in `tt_metal/hw/inc/api/compute/eltwise_unary/typecast.h` reach the same function, so Float32, Float16_b, Bfp8_b and Bfp4_b sources are all affected: lines 122, 156, 188 and 309.

The kernel has no `DISABLE_SFPLOADMACRO` variant, so this is the only code path.

### How it was verified

ttsim Blackhole (`TT_METAL_SIMULATOR=libttsim_bh.so`, slow dispatch), driving the boundary values and a uniform random bit-pattern sweep through `calculate_typecast_fp32_to_int32`, compared against `torch.Tensor.to(torch.int32)`.

The tests do not reach the overflow. `tests/ttnn/python_api_testing/typecast_test_helpers.py:38` bounds inputs to `(-1000, 80000)` via `_CLAMP_HIGH = _DTYPE_MAX[ttnn.uint16] + 14465`, and `tt_metal/tt-llk/tests/python_tests/test_eltwise_unary_typecast.py:179` feeds whole numbers through `_whole_number_float_spec(16 if bfp_involved else 201)`.

### Validation, measured before and after

Branch `fix/typecast-int32-saturate` on `seonglae/tt-metal`, commit `283ab1c1`: two kernel files and one test, 102 insertions.

The change is the saturation constant. `ckernel_sfpu_typecast.h` loaded INT_MIN as the single constant for both signs; it now loads INT_MAX, in two halves because the `FLOATB` immediate only reaches the upper 16 bits. The existing two's-complement negate then produces `-INT_MAX` for a negative overflow, so both ends of the range land where **this repository's own golden already puts them**: `saturate_integer` in `golden_generators.py` clamps a signed destination to `[iinfo.min + 1, iinfo.max]`, and its comment says why, *"+1 because hardware uses sign-magnitude representation"*. Nothing else in the sequence moves and the in-range path is untouched.

Measured on ttsim Blackhole, reading int32 back through the two's-complement decoder:

| input | before | after | golden |
|---|---|---|---|
| `2^31` | -2147483648 | **2147483647** | 2147483647 |
| `3e9` | -2147483648 | **2147483647** | 2147483647 |
| `3.4e38` | -2147483648 | **2147483647** | 2147483647 |
| `+inf` | -2147483648 | **2147483647** | 2147483647 |
| `-2^31` | -2147483648 | **-2147483647** | -2147483647 |
| `-inf` | -2147483648 | **-2147483647** | -2147483647 |
| `2147483520.0` | 2147483520 | 2147483520 | unchanged |
| `-100`, `-2e9`, `-1`, `0`, `+/-0.5`, `100` | exact | exact | unchanged |

Three tests, one per behaviour. Against the unpatched kernel **the two overflow tests fail and the in-range test passes**, which is why the third one is there: it shows the change is confined to the overflow lanes.

`NaN` stays divergent and is deliberately out of scope. The kernel folds it into the same overflow lane, so it now reads `2147483647` where torch gives `0`. Separating it costs an exponent and a mantissa test inside the hot loop, and it is a semantics question rather than a saturation one, so it wants a decision before the instructions are spent.

### The one decision left

Matching torch exactly, rather than the golden, would need `INT32_MIN` at the negative end instead of `-INT32_MAX`, and that needs a second constant selected on the sign of the input before the exponent mask narrows. The awkward part is that the two's-complement negate currently doubles as the sign application, so the positive-overflow lane would need its own predicate rather than a negation of the existing constant, which costs instructions in the hot loop.

The patch above takes the golden's answer instead, because the golden states the reason for it. Which of the two is the contract is yours to settle, and it is the same question as the NaN arm.

## Prior art

**#51304** (open) quotes this exact function from this exact file and scopes the question out in writing: *"Note this issue is only about truncation vs round-to-nearest-even. The composite path also does not saturate to the int8 range the way the QUANT LLK does; that is a separate behavioural question and is deliberately out of scope here."* So the saturation behaviour is known to be unexamined rather than settled.

**#51655** (open, `[Bounty $1000]`) names `ckernel_sfpu_typecast.h:183` in a table row reading `| float -> int32 | truncates |`, characterising the line as correct on the rounding axis. **#55325** is integer-to-integer narrowing in `calculate_typecast_uint32_to_uint16` and states that `uint32 -> int32` is a widening move and unaffected. **#53639**, the fix PR for #51655, touches only the five uint16 sites and reports that *"int32, uint32 and uint8 stay at 0"*.

Four open PRs touch this file and none fixes it: #53639 (uint16 only), #54351 (draft, dest-acc threading), #55486 (adds int8 paths, leaving `calculate_typecast_fp32_to_int32` untouched), and **#55423**, a CRTP dispatch refactor that wraps the kernel verbatim in a `Typecast<>` struct without altering the body.

Full-text searches for `typecast int32 saturate`, `INT_MIN typecast`, `int32 saturation` and `typecast overflow int32` return nothing on this defect. Every `typecast in:title` hit is about uint16 rounding or uint8-versus-uint16 narrowing: #55486, #55325, #53639, #51655, #51227.

Two closed issues are adjacent and neither covers it. **#55502**, `[Bounty $5000] Fix INT_MIN correctness in int32 div, remainder, fmod, and scalar promotion`, is the same class in different ops. **#33976**, `[Bounty $12500] Optimise/improve rounding/typecast LLKs`, has `fp32/bf16 to i32/u32 (done)` on its checklist, so this kernel already had a paid pass and the saturation constant survived it.

## Acceptance criteria

1. `ttnn.typecast(t, ttnn.int32)` returns `2147483647` for every finite float at or above `2^31` and for `+inf`, `-2147483648` for `-inf` and for finite floats at or below the negative bound, and `0` for NaN.
2. The negative side and fractional truncation are unchanged: `-2147483648.0` still gives `INT32_MIN`, and in-range fractions still truncate toward zero.
3. All four source dtypes behave identically, since `typecast.h:152`, `:118`, `:184` and `:305` share the function.
4. The `int32` arm of `_typecast_golden_function` states its bound explicitly, the way the uint8, uint16 and uint32 arms already do, rather than inheriting it from a host cast.
5. `tt_metal/tt-llk/tests/python_tests/test_eltwise_unary_typecast.py` covers `2^31 - 128`, `2^31`, `3e9`, `FLT_MAX`, `±inf`, `NaN` and `-2^31`, none of which its current `[0, 201)` stimuli reach.
6. Whether the negative bound is `INT32_MIN` or `-(2^31 - 1)` is settled and documented, since the tt-llk golden clamps to the latter on sign-magnitude grounds.

## Reproducing on ttsim, no silicon needed

The toolchain is hash-pinned by the repo itself at `tt_metal/tt-llk/tests/sfpi-version`, so no silicon and no separate toolchain are needed.

The patch ships the reproducer as `tt_metal/tt-llk/tests/python_tests/test_typecast_int32_saturation.py`. **Take that file alone, leave the kernel unpatched, and two of its three tests fail**, which is the defect. The third asserts the in-range and boundary values and passes either way, which is what shows the change is confined to the overflow lanes.

```bash
cd tt_metal/tt-llk/tests/python_tests
env TT_METAL_SIMULATOR=/path/to/libttsim_bh.so TT_METAL_DISABLE_SFPLOADMACRO=1 \
    python -m pytest -q -s --run-simulator test_typecast_int32_saturation.py
```

Unpatched: `test_positive_overflow_clamps_to_int32_max` reports `[-2147483648] * 4` against `[2147483647] * 4`, and the negative test reports the same left-hand side against `[-2147483647] * 4`. Patched: three passed.

`twos_complement=True` on the `StimuliConfig` is load-bearing and easy to lose, since it defaults to `False`. Without it `unpack_int32` decodes the two's-complement result as sign-magnitude, which turns `-100` into `-2147483548` and `INT32_MIN` into `0`, and a reproducer can spend a while attributing that to the kernel. `tests/helpers/include/sfpu_operations.h:2112` records the same constraint from the other side. `TT_METAL_DISABLE_SFPLOADMACRO=1` does not change this kernel's arithmetic; there is no macro variant of it.

**Limitations of what I ran.** Blackhole only, on the simulator, which is bit-exact for arithmetic and not cycle-accurate. I have no silicon. Wormhole not measured, as noted above.

**Comment by MyDude92** [comment]
Hi @tenstorrent team,

I would like to claim and execute the fix for **#55457** to eliminate the `-inf` divergence in `ttnn.log_sigmoid` for large positive inputs, restoring correct asymptotic convergence to `0.0`.

I have already implemented and verified the piecewise formulation with a passing unit test suite matching the issue's failing data table.

---

### <u>**Technical Remediation Plan**</u>

#### **1. Problem & Divergence Analysis**
The mathematical definition of log_sigmoid is:
* `log_sigmoid(x) = log(1 / (1 + exp(-x))) = -log(1 + exp(-x))`
* As `x -> +inf`, `sigmoid(x) -> 1.0`, meaning `log_sigmoid(x)` **must asymptotically approach 0.0**.

On hardware, inputs `x > 172` in bfloat16 trigger overflow in the large-positive branch of `ckernel_sfpu_logsigmoid.h`, causing the output to explode to `-5.7e34` and eventually diverge to `-inf`.

#### **2. Proposed Formulation**
We implement a stable piecewise operator:
* **For `x >= 20.0`**: Clamp strictly to `0.0` (in bfloat16 and float32, `exp(-20) < 2e-9` is below machine precision, making `1 + exp(-x)` identically `1.0`, and `log(1.0) = 0.0`).
* **For `0 <= x < 20.0`**: Compute `-log1p(exp(-x))`. Since `-x <= 0`, `exp(-x)` is bounded in `(0, 1]`, preventing overflow.
* **For `x < 0.0`**: Compute `x - log1p(exp(x))`. Since `x < 0`, `exp(x)` is bounded in `(0, 1]`, preventing overflow.

#### **3. Verification Results (Matching Issue Table)**

| Input Value (bfloat16) | Broken Hardware Output | Solved Stable Output | True Target | Status |
| :--- | :--- | :--- | :--- | :--- |
| **x = 170.0** | -0.0000 | **0.000000** | 0.000000 | **PASS** |
| **x = 172.0** | -0.0185 (Drifting) | **0.000000** | 0.000000 | **PASS** |
| **x = 174.0** | -0.1337 (Drifting) | **0.000000** | 0.000000 | **PASS** |
| **x = 176.0** | -0.9765 (Drifting) | **0.000000** | 0.000000 | **PASS** |
| **x = 256.0** | -5.71153e+34 (Explosion) | **0.000000** | 0.000000 | **PASS** |
| **x = 266.0** | -inf (Divergence) | **0.000000** | 0.000000 | **PASS** |
| **x = 0.0** | -0.6931 | **-0.693147** | -ln(2) | **PASS** |
| **x = -1.0** | -1.3132 | **-1.313262** | -1.313262 | **PASS** |
| **x = -50.0** | -50.0000 | **-50.000000** | -50.000000 | **PASS** |

---

### <u>**Deliverable Handover Package**</u>
* **Implementation Code**: https://github.com/MyDude92/keep/blob/main/bounties/bounty_12_log_sigmoid_stable.py
* **Passing Test Suite**: `tests/test_log_sigmoid_bounty.py` (**100% Passing**)
* **Full Audit Dossier**: https://github.com/MyDude92/keep/blob/main/bounties/BOUNTY_12_LOG_SIGMOID_DOSSIER.md

Could you please assign this issue to me so I can submit the PR?

Thanks,  
**Alistair** / `@MyDude92`

**Comment by jdymitarai** [comment]
Hi @tenstorrent team,

I would like to claim and resolve this issue (#55933) to correct the positive float-to-int32 overflow saturation in `ttnn.typecast`.

### Proposed Remediation Plan:
1. **SFPU Kernel Saturation (`ckernel_sfpu_typecast.h`)**:
   - In `calculate_typecast_fp32_to_int32`, fix the positive overflow lanes where inputs $|x| \ge 2^{31}$ (and `+inf`) incorrectly retain the initial single saturation constant (`0x80000000` / `INT32_MIN`).
   - Clamp positive overflow cleanly to `INT32_MAX` (`0x7FFFFFFF` / `2147483647`), while preserving negative saturation at `INT32_MIN` (`-2147483648`) and in-range fractional truncation toward zero.
   - Ensure the fix applies consistently across both Wormhole B0 and Blackhole architectures.
2. **Golden & Typecast Semantics**:
   - Align `_typecast_golden_function` / `integer_golden` bounds with device saturation contracts.
   - Explicitly document and handle boundary semantics (`2^31 - 128`, `2^31`, `+inf`, `-inf`, `NaN`).
3. **Regression Test Suite**:
   - Add targeted test cases to `tests/ttnn/python_api_testing/` and `tt_metal/tt-llk/tests/python_tests/test_eltwise_unary_typecast.py` covering extreme positive values (`2^31`, `3e9`, `1e38`, `+inf`) and boundary checks to prevent future sign-inversion regressions.

Could you please assign this issue to me so I can proceed with submitting the validated PR? Thank you!
